### PR TITLE
feature: use binary keys instead of passwords

### LINKED ISSUE #52: Use binary keys to protect data
In order to support "FxA as Master Password" (mozilla-lockbox/lockbox-extension#209), the datastore needs to support encryption using a binary key.

* When `initialize()` is called, allow the API user to pass in a JWK or Uint8Array that represents a symmetric key in order to prepare the databases.
   - Allow the API user to call `initialize()` without a key, and use a consistent "null" key value to support the low-touch "quickstart"
* When `unlock()` is called, allow the API user to pass in a JWK or Uint8Array that represents the symmetric key used when initializing.
   - Allow the API user to call `unlock()` without a key, and use the default "null" key value

### PR DESCRIPTION
**Pull request by linuxwolf**
Resolves #52 

Switches the datastore from using a password-based key to using a binary key (from FxA).

This work is assuming there is no migration from the former model to this new model.  It is also providing a path to incorporating an external salt for key derivations; in practice this will be the FxA uid, or a persisted UUID while the system is not bound to an FxA account.

### CODE REVIEW

**Comment by devinreams** [comment]
@linuxwolf for purposes of releasing https://github.com/mozilla-lockbox/lockbox-extension/pull/362 would you like this reviewed by @jimporter (knowing a broader review is needed from others soon)?

Or do you want to keep this branch and keep the extension pegged to it for now?

**Comment by linuxwolf** [comment]
@jimporter @devinreams This PR should be good to review now.



> Or do you want to keep this branch and keep the extension pegged to it for now?

I don't think it's worth keeping separate; does mean PR should land before mozilla-lockbox/lockbox-extension#362 (and that PR updated to point to the shared repo).

File: lib/datastore.js [hunk 1]
```diff
@@ -135,47 +167,64 @@ class DataStore {
...
    *
    * @param {Object} opts The initialization options
-   * @param {String} [opts.password=""] The master password to lock with
-   * @param {String} [opts.salt] The salt to use in deriving the master
-   *        key.
-   * @param {Number} [opts.iterations=8192] The iteration count to use in
-   *        deriving the master key
+   * @param {jose.JWK.Key} [opts.appKey] The master app key to setup with
+   * @param {string} [opts.salt] The salt to use in deriving the master
+   *        keys.
```
Comment (jimporter) [inline lib/datastore.js]: Micro-nit: no punctuation seems to be the consensus here.

File: lib/itemkeystore.js [hunk 2]
```diff
@@ -41,33 +33,36 @@ class ItemKeyStore {
...
   /**
    * The group name.
    *
-   * @type {String}
+   * @type {string}
    * @readonly
    */
   get group() { return instance.get(this).group || ""; }
   /**
+<<<<<<< HEAD
```
Comment (jimporter) [inline lib/itemkeystore.js]: Looks like a merge conflict didn't get resolved here.

File: lib/itemkeystore.js [hunk 3]
```diff
@@ -77,15 +72,20 @@ class ItemKeyStore {
    * a plain Object with the following properties:
    *
    * * `group` {**String**} The name of the keystore's group
+<<<<<<< HEAD
```
Comment (jimporter) [inline lib/itemkeystore.js]: Another merge conflict here.

File: .eslintrc.json [hunk 4]
```diff
@@ -1,20 +1,37 @@
...
   "root": true,
   "rules": {
+    "jsdoc/check-types": "error",
+    "security/detect-non-literal-fs-filename": "off",
+    "security/detect-object-injection": "off",
+
+    "eqeqeq": "error",
     "indent": ["error", 2, {"SwitchCase": 1, "VariableDeclarator": {"var": 2, "let": 2, "const": 3}}],
     "linebreak-style": ["error", "unix"],
+    "no-var": "off",
```
Comment (jimporter) [inline .eslintrc.json]: What do we need `var` for? I only see it in a couple places in this diff...

File: lib/datastore.js [hunk 5]
```diff
@@ -84,7 +114,8 @@ class DataStore {
    * See {@link datastore.create} for the details of what `{cfg}`
    * parameters are supported.
    *
-   * @param {Onject} cfg - The configuration parameters
+   * @param {Object} cfg - The configuration parameters.
```
Comment (jimporter) [inline lib/datastore.js]: Nit: The consensus seems to be no hyphen and no punctuation. I don't have a preference for how we do this, but consistency would be nice.

File: .eslintrc.json (hunk 4 above)
Comment (linuxwolf) [reply .eslintrc.json]: `var ` is used in the build-utils files, which are also linted.

Comment (jimporter) [reply .eslintrc.json]: Does it need to be `var` there, and if so, would it be better to just ignore the `no-var` rule on those lines?

Comment (linuxwolf) [reply .eslintrc.json]: node.js still has some issues with `let` at the top-level.  I can go back through those files to fix it, but really rather file a chore for now.

**Pull request merged**